qualtran/bloqs/arithmetic/subtraction.py

```diff
@@ -167,3 +167,84 @@ def _sub_diff_size_regs() -> Subtract:
 _SUB_DOC = BloqDocSpec(
     bloq_cls=Subtract, examples=[_sub_symb, _sub_small, _sub_large, _sub_diff_size_regs]
 )
+
+
+@frozen
+class SubtractFrom(Bloq):
+    """A version of `Subtract` that leaves behind the value being subtracted.
+
+    Implements $U|a\rangle|b\rangle \rightarrow |a\rangle|b - a\rangle$, essentially equivalent to
+    the statement `b -= a`.
+
+    Args:
+        dtype: Quantum datatype used to represent the integers a, b, and a - b.
+
+    Registers:
+        a: A dtype.bitsize-sized input register (register a above).
+        b: A dtype.bitsize-sized input/output register (register b above).
+    """
+
+    dtype: Union[QInt, QUInt, QMontgomeryUInt] = field()
+
+    @dtype.validator
+    def _dtype_validate(self, field, val):
+        if not isinstance(val, (QInt, QUInt, QMontgomeryUInt)):
+            raise ValueError("Only QInt, QUInt and QMontgomeryUInt types are supported.")
+
+    @property
+    def signature(self):
+        return Signature([Register("a", self.dtype), Register("b", self.dtype)])
+
+    def on_classical_vals(
+        self, a: 'ClassicalValT', b: 'ClassicalValT'
+    ) -> Dict[str, 'ClassicalValT']:
+        unsigned = isinstance(self.dtype, (QUInt, QMontgomeryUInt))
+        bitsize = self.dtype.bitsize
+        N = 2**bitsize if unsigned else 2 ** (bitsize - 1)
+        return {'a': a, 'b': int(math.fmod(b - a, N))}
+
+    def wire_symbol(
+        self, reg: Optional['Register'], idx: Tuple[int, ...] = tuple()
+    ) -> 'WireSymbol':
+        from qualtran.drawing import directional_text_box
+
+        if reg is None:
+            return Text('')
+        if reg.name == 'a':
+            return directional_text_box('a', side=reg.side)
+        elif reg.name == 'b':
+            return directional_text_box('b - a', side=reg.side)
+        else:
+            raise ValueError()
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        return {(Negate(self.dtype), 1), (Subtract(self.dtype, self.dtype), 1)}
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', a: Soquet, b: Soquet) -> Dict[str, 'SoquetT']:
+        a, b = bb.add_t(Subtract(self.dtype, self.dtype), a=a, b=b)  # a, a - b
+        b = bb.add(Negate(self.dtype), x=b)  # a, b - a
```

Comment:
This effectively reduces to `Negate(b); Add(a, b); Negate(b)` right? I think it would be cheaper to use ones-complement (`BitwiseNot`) instead of negate, because `Negate(a) = BitwiseNot(a); AddK(a, 1)` and we can avoid the `AddK` which requires Toffolis.
```
BitwiseNot(b) # a, -1 - b
Add(a, b) # a, a - 1 - b
BitwiseNot(b) # a, -1 - (a - 1 - b) = a, -a + b
```

But this isn't blocking this PR. I was going to add a `BitwiseNot` gate, I can update this decomposition then.

Reply:
Sounds great, let's switch to `BitwiseNot` in the next PR.